pyro/util.py

```diff
@@ -311,6 +312,8 @@ def check_model_guide_match(model_trace, guide_trace):
     also appears in the model, (3) at each sample site that appears in both
     the model and guide, the model and guide agree on sample shape.
     """
+    if not pyro.infer.is_validation_enabled():
```

Comment:
ditto: It seems cleaner to gate outside the `check_*()` functions

Reply:
My only (minor) quibble with that is that we will need to remember to gate checks inside each of the `trace*_ELBO` classes any time we are using these checks, and for any subsequent checks that we add. e.g. (note the four checks here)

```
if infer.validation_enabled():
    check_model_guide_match(model_trace, guide_trace)
guide_trace = prune_subsample_sites(guide_trace)
model_trace = prune_subsample_sites(model_trace)
if infer.validation_enabled():
    check_traceenum_requirements(model_trace, guide_trace)

...
if infer.validation_enabled():
    check_site_shape(site, self.max_iarange_nesting)
...
if infer.validation_enabled():
    check_site_shape(site, self.max_iarange_nesting)
```

Centralizing this then seems more cleaner, at the cost of a little less control and some additional looping (which should be negligible cost for all practical purposes). I don't have a strong opinion on this though.

Reply:
I do feel strongly that `check_site_site()` should check the site's shape, rather than *maybe* check the site's shape, depending on some global state. E.g. I might want to call `check_site_shape()` in a debugger or in a custom inference algorithm, and in those use case I don't want to think about global state.

Reply:
> I do feel strongly that check_site_site() should check the site's shape, rather than maybe check the site's shape

That's a fair point.